File: hiroo-js/src/bot.js
import { HTTPClient } from "./http.js";
import { GatewayClient } from "./gateway.js";
import { defaultIntents } from "./intents.js";
import { connectVoice } from "./voice.js";

/**
 * CommandContext passed to slash command / interaction handlers.
 *
 * - `respond(content, {embeds, components, ephemeral})` — for interactions
 *   uses the interaction callback endpoint, otherwise falls back to
 *   posting a regular message in the source channel.
 * - `defer()` — tell the user "thinking…" (only for interactions).
 * - `followup(...)` — send a follow-up message after deferring.
 */
export class CommandContext {
  constructor(bot, { message = null, interaction = null, args = {} } = {}) {
    this.bot = bot;
    this.message = message;
    this.interaction = interaction;
    this.args = args;
    if (interaction) {
      this.channelId = interaction.channel_id ?? null;
      this.dmId = interaction.dm_id ?? null;
      this.guildId = interaction.guild_id ?? null;
      this.interactionId = interaction.id ?? null;
    } else if (message) {
      this.channelId = message.channel_id ?? null;
      this.dmId = message.dm_id ?? null;
      this.guildId = message.server_id ?? message.guild_id ?? null;
      this.interactionId = null;
    } else {
      this.channelId = this.dmId = this.guildId = this.interactionId = null;
    }
  }

  get author() {
    if (this.message) return this.message.author ?? null;
    if (this.interaction) return this.interaction.user ?? null;
    return null;
  }

  async respond(content = "", { embeds, components, ephemeral = false } = {}) {
    if (this.interactionId) {
      return this.bot.http.interactionCallback(this.interactionId, {
        type: 4, content, ephemeral, embeds, components,
      });
    }
    return this.bot.http.sendMessage({
      channelId: this.channelId, dmId: this.dmId, content, embeds, components,
    });
  }

  async defer() {
    if (!this.interactionId) return;
    await this.bot.http.interactionCallback(this.interactionId, { type: 5 });
  }

  async followup(content = "", opts = {}) {
    if (!this.interactionId) return this.respond(content, opts);
    return this.bot.http.interactionFollowup(this.interactionId, { content, ...opts });
  }
}

const TYPE_MAP = { string: 3, number: 4, boolean: 5, float: 10 };

/**
 * Main entry point for a HiRoo bot.
 *
 *   const bot = new Bot({ intents: defaultIntents() });
 *   bot.on("ready", () => console.log("ready"));
 *   bot.on("message_create", async (msg) => { ... });
 *   bot.slashCommand({ name: "echo", description: "Echo" }, async (ctx, { text }) => {
 *     await ctx.respond(text);
 *   }, [{ name: "text", type: "string", required: true }]);
 *   await bot.run(process.env.BOT_TOKEN);
 */
export class Bot {
  constructor({
    intents = defaultIntents(),
    baseURL = "https://hiroo.intave.tech",
    applicationId = null,
    shardCount = 1,
  } = {}) {
    this.intents = Number(intents);
    this.baseURL = baseURL;
    this.applicationId = applicationId;
    this.shardCount = shardCount;
    this.user = null;
    this.http = new HTTPClient({ token: "", baseURL });
    this._handlers = new Map();    // event name → array of handlers
    this._commands = new Map();    // command name → { handler, options, ... }
    this._componentHandlers = new Map();
    this._shards = [];
    this._voiceConnections = [];
  }

  /** Подключиться к голосовому каналу (или DM-звонку). */
  async connectVoice({ channelId, dmId } = {}) {
    return connectVoice(this, { channelId, dmId });
  }

  /** Зарегистрировать VoiceConnection, чтобы run() корректно отключил его по SIGINT. */
  _registerVoice(vc) { this._voiceConnections.push(vc); }

  /**
   * Register an event handler. Discord-style short names are mapped to
   * gateway event names (message → message_create, reaction → reaction_add,
   * etc.) automatically.
   */
  on(event, handler) {
    const aliases = {
      message: "message_create",
      message_edit: "message_update",
      reaction: "reaction_add",
      member_join: "guild_member_add",
      member_leave: "guild_member_remove",
      voice_state: "voice_state_update",
      presence: "presence_update",
    };
    const resolved = aliases[event] ?? event;
    if (!this._handlers.has(resolved)) this._handlers.set(resolved, []);
    this._handlers.get(resolved).push(handler);
    return this;
  }

  /**
   * Register a slash command.
   *
   *   bot.slashCommand({ name: "add", description: "Sum" },
   *     async (ctx, { a, b }) => ctx.respond(String(a + b)),
   *     [{ name: "a", type: "number", required: true },
   *      { name: "b", type: "number", required: true }]);
   */
  slashCommand({ name, description = "", guildId = null }, handler, options = []) {
    const normalized = options.map((o) => ({
      name: o.name,
      description: o.description ?? o.name,
      type: typeof o.type === "number" ? o.type : (TYPE_MAP[o.type] ?? 3),
      required: o.required ?? false,
    }));
    this._commands.set(name, {
      handler,
      type: "slash",
      name,
      description: description || name,
      options: normalized,
      guildId,
    });
    return this;
  }

  /** Register a handler for a component interaction (button/select). */
  component(customId, handler) {
    this._componentHandlers.set(customId, handler);
    return this;
  }

  async _dispatch(event, data) {
    if (event === "ready") {
      this.user = data.user ?? null;
      const shardArr = data.shard ?? [0, 1];
      this.shardCount = shardArr[1];
      if (this.applicationId) {
        for (const cmd of this._commands.values()) {
          try {
            await this.http.registerCommand(this.applicationId, cmd);
          } catch (e) {
            console.error(`[hiroo] failed to register /${cmd.name}:`, e?.message ?? e);
          }
        }
      }
    }

    // Decorate message_create with a .reply helper.
    let payload = data;
    if (event === "message_create") {
      payload = {
        ...data,
        reply: async (content = "", opts = {}) =>
          this.http.sendMessage({
            channelId: data.channel_id, dmId: data.dm_id,
            content, ...opts, replyToId: data.id,
          }),
      };
      // Auto-invoke slash command if content starts with `/<name>`.
      const content = data.content ?? "";
      if (content.startsWith("/")) {
        const first = content.slice(1).split(/\s+/, 1)[0];
        const cmd = this._commands.get(first);
        if (cmd && !(data.author?.bot)) {
          const rest = content.slice(first.length + 1).trim();
          const parts = rest ? rest.split(/\s+/) : [];
          const kwargs = {};
          for (let i = 0; i < cmd.options.length; i++) {
            const o = cmd.options[i];
            // The last option soaks up the remaining whitespace.
            const raw = i === cmd.options.length - 1
              ? parts.slice(i).join(" ")
              : parts[i];
            if (raw == null || raw === "") continue;
            kwargs[o.name] = coerceArg(raw, o.type);
          }
          const ctx = new CommandContext(this, { message: data, args: kwargs });
          try { await cmd.handler(ctx, kwargs); }
          catch (e) { console.error(`[hiroo] command /${first} failed:`, e); }
        }
      }
    }

    if (event === "interaction_create") {
      const inter = data ?? {};
      if (inter.type === "component") {
        const h = this._componentHandlers.get(inter.custom_id ?? "");
        if (h) {
          const ctx = new CommandContext(this, { interaction: inter, args: inter.data ?? {} });
          try { await h(ctx); }
          catch (e) { console.error(`[hiroo] component ${inter.custom_id} failed:`, e); }
        }
      } else if (inter.type === "command") {
        const cmd = this._commands.get(inter.command_name ?? "");
        if (cmd) {
          const opts = (inter.data ?? {}).options ?? {};
          const ctx = new CommandContext(this, { interaction: inter, args: opts });
          try { await cmd.handler(ctx, opts); }
          catch (e) { console.error(`[hiroo] slash /${inter.command_name} failed:`, e); }
        }
      }
    }

    const handlers = this._handlers.get(event) ?? [];
    for (const h of handlers) {
      try { await h(payload); }
      catch (e) { console.error(`[hiroo] handler for ${event} failed:`, e); }
    }
  }

  /** Connect all shards. Resolves when every shard's reconnect loop exits. */
  async run(token) {
    this.http.token = token;
    this._shards = Array.from({ length: this.shardCount }, (_, i) =>
      new GatewayClient({
        url: this.baseURL, token, intents: this.intents,
        shardId: i, shardCount: this.shardCount,
        onDispatch: (event, data) => this._dispatch(event, data),
      })
    );
    const stop = async () => {
      for (const vc of this._voiceConnections) {
        try { await vc.disconnect(); } catch {}
      }
      this._voiceConnections.length = 0;
      try { await this.http.request("DELETE", "/api/voice/presence"); } catch {}
      for (const s of this._shards) { try { await s.close(); } catch {} }
      process.exit(0);
    };
    process.once("SIGINT", stop);
    process.once("SIGTERM", stop);
    await Promise.all(this._shards.map((s) => s.connect()));
  }
}

function coerceArg(raw, type) {
  // type is a numeric APPLICATION_COMMAND_OPTION_TYPE; see TYPE_MAP.
  switch (type) {
    case 4: { const n = parseInt(raw, 10); return Number.isFinite(n) ? n : raw; }
    case 5: return /^(1|true|yes|y)$/i.test(raw);
    case 10: { const f = parseFloat(raw); return Number.isFinite(f) ? f : raw; }
    default: return raw;
  }
}

File: hiroo-java/src/main/java/tech/intave/hiroo/Bot.java
package tech.intave.hiroo;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.function.BiConsumer;

/**
 * Точка входа в HiRoo-бота.
 *
 * <pre>{@code
 * Bot bot = new Bot.Builder()
 *     .intents(Intents.defaults() | Intents.MESSAGE_CONTENT)
 *     .applicationId(System.getenv("APP_ID"))
 *     .build();
 *
 * bot.on("ready", (evt, data) ->
 *     System.out.println("Logged in as " + data.path("user").path("username").asText()));
 *
 * bot.slashCommand("echo", "Echoes back your text",
 *     List.of(SlashOption.string("text", "Any text", true)),
 *     (ctx, args) -> ctx.respond(args.path("text").asText()));
 *
 * bot.run("Bot " + System.getenv("BOT_TOKEN"));
 * }</pre>
 */
public class Bot {
    private final int intents;
    private final String baseUrl;
    private final String applicationId;
    private final int shardCount;
    private final HttpClient http;
    private final ObjectMapper mapper = new ObjectMapper();

    private final Map<String, List<BiConsumer<String, JsonNode>>> handlers = new HashMap<>();
    private final Map<String, SlashRegistration> commands = new LinkedHashMap<>();
    private final List<Gateway> shards = new CopyOnWriteArrayList<>();

    public Bot(Builder b) {
        this.intents = b.intents;
        this.baseUrl = b.baseUrl;
        this.applicationId = b.applicationId;
        this.shardCount = Math.max(1, b.shardCount);
        this.http = new HttpClient("", this.baseUrl);
    }

    public HttpClient http() { return http; }
    public String applicationId() { return applicationId; }

    /** Регистрирует обработчик события шлюза. Несколько обработчиков на событие допустимо. */
    public Bot on(String event, BiConsumer<String, JsonNode> handler) {
        handlers.computeIfAbsent(event, k -> new ArrayList<>()).add(handler);
        return this;
    }

    /**
     * Регистрирует слэш-команду. Авто-регистрируется на сервере при получении
     * события {@code ready}, если задан {@code applicationId}. Плюс обрабатывает
     * как интеракции (через пикер), так и текстовые вызовы {@code /name args…}.
     */
    public Bot slashCommand(String name, String description, List<SlashOption> options, SlashHandler handler) {
        return slashCommand(name, description, options, handler, null);
    }

    public Bot slashCommand(String name, String description, List<SlashOption> options,
                            SlashHandler handler, String guildId) {
        commands.put(name, new SlashRegistration(name, description,
            options == null ? Collections.emptyList() : options, handler, guildId));
        return this;
    }

    /** Блокирующий запуск: поднимает шарды и возвращает управление только после их остановки. */
    public void run(String token) throws InterruptedException {
        http.setToken(token);
        List<Thread> threads = new ArrayList<>();
        for (int i = 0; i < shardCount; i++) {
            final int shardId = i;
            Gateway g = new Gateway(baseUrl, token, intents, shardId, shardCount, this::dispatch);
            shards.add(g);
            Thread t = new Thread(() -> {
                try { g.run(); }
                catch (InterruptedException ignored) { Thread.currentThread().interrupt(); }
            }, "hiroo-shard-" + shardId);
            threads.add(t);
            t.start();
        }
        Runtime.getRuntime().addShutdownHook(new Thread(this::shutdown, "hiroo-shutdown"));
        for (Thread t : threads) t.join();
    }

    public void shutdown() {
        // Снять голосовое присутствие на бэкенде — эндпоинт идемпотентный.
        try { Voice.unregisterPresence(this); } catch (Exception ignored) {}
        for (Gateway g : shards) { try { g.close(); } catch (Exception ignored) {} }
    }

    // ── Диспатч ──────────────────────────────────────────────────────

    private void dispatch(String event, JsonNode data) {
        // 1) Авто-регистрация команд на ready.
        if ("ready".equals(event) && applicationId != null && !commands.isEmpty()) {
            for (SlashRegistration cmd : commands.values()) {
                try {
                    List<Map<String, Object>> opts = new ArrayList<>();
                    for (SlashOption o : cmd.options) opts.add(o.toMap());
                    http.registerCommand(applicationId, "slash", cmd.name, cmd.description, opts, cmd.guildId);
                } catch (Exception e) {
                    System.err.println("[hiroo] failed to register /" + cmd.name + ": " + e.getMessage());
                }
            }
        }

        // 2) Интеракции (кнопки, слэш из пикера).
        if ("interaction_create".equals(event)) {
            String itype = data.path("type").asText("");
            if ("command".equals(itype)) {
                String name = data.path("command_name").asText("");
                SlashRegistration cmd = commands.get(name);
                if (cmd != null) {
                    JsonNode args = data.path("data").path("options");
                    if (args.isMissingNode() || args.isNull()) args = mapper.createObjectNode();
                    CommandContext ctx = CommandContext.fromInteraction(this, data, args);
                    try { cmd.handler.handle(ctx, args); }
                    catch (Exception e) { System.err.println("[hiroo] slash /" + name + " failed: " + e); }
                }
            }
        }

        // 3) Текстовый вызов /<name> …
        if ("message_create".equals(event)) {
            String content = data.path("content").asText("");
            boolean isBot = data.path("author").path("bot").asBoolean(false);
            if (!isBot && content.startsWith("/")) {
                int firstSpace = content.indexOf(' ');
                String first = firstSpace < 0 ? content.substring(1) : content.substring(1, firstSpace);
                SlashRegistration cmd = commands.get(first);
                if (cmd != null) {
                    String argText = firstSpace < 0 ? "" : content.substring(firstSpace + 1).trim();
                    ObjectNode args = buildArgsFromText(cmd, argText);
                    CommandContext ctx = CommandContext.fromMessage(this, data, args);
                    try { cmd.handler.handle(ctx, args); }
                    catch (Exception e) { System.err.println("[hiroo] text /" + first + " failed: " + e); }
                }
            }
        }

        // 4) Пользовательские подписчики.
        List<BiConsumer<String, JsonNode>> list = handlers.get(event);
        if (list != null) {
            for (BiConsumer<String, JsonNode> h : list) {
                try { h.accept(event, data); }
                catch (Exception e) { System.err.println("[hiroo] handler for " + event + " failed: " + e); }
            }
        }
    }

    /** Простой парсер argv: слова через пробел, последняя опция забирает хвост. */
    private ObjectNode buildArgsFromText(SlashRegistration cmd, String argText) {
        ObjectNode args = mapper.createObjectNode();
        if (cmd.options.isEmpty() || argText.isEmpty()) return args;
        String[] parts = argText.split("\\s+", -1);
        for (int i = 0; i < cmd.options.size(); i++) {
            SlashOption o = cmd.options.get(i);
            String raw;
            if (i >= parts.length) break;
            if (i == cmd.options.size() - 1) {
                StringBuilder sb = new StringBuilder();
                for (int j = i; j < parts.length; j++) {
                    if (sb.length() > 0) sb.append(' ');
                    sb.append(parts[j]);
                }
                raw = sb.toString();
            } else {
                raw = parts[i];
            }
            if (raw == null || raw.isEmpty()) continue;
            switch (o.type) {
                case INTEGER:
                    try { args.put(o.name, Long.parseLong(raw)); }
                    catch (NumberFormatException ignored) { args.put(o.name, raw); }
                    break;
                case BOOLEAN:
                    args.put(o.name, raw.matches("(?i)^(1|true|yes|y)$"));
                    break;
                case NUMBER:
                    try { args.put(o.name, Double.parseDouble(raw)); }
                    catch (NumberFormatException ignored) { args.put(o.name, raw); }
                    break;
                default:
                    args.put(o.name, raw);
            }
        }
        return args;
    }

    // ── Внутренняя структура ─────────────────────────────────────────

    private static final class SlashRegistration {
        final String name;
        final String description;
        final List<SlashOption> options;
        final SlashHandler handler;
        final String guildId;
        SlashRegistration(String name, String description, List<SlashOption> options,
                          SlashHandler handler, String guildId) {
            this.name = name;
            this.description = description;
            this.options = options;
            this.handler = handler;
            this.guildId = guildId;
        }
    }

    // ── Builder ──────────────────────────────────────────────────────

    public static class Builder {
        private int intents = Intents.defaults();
        private String baseUrl = "https://hiroo.intave.tech";
        private String applicationId;
        private int shardCount = 1;

        public Builder intents(int intents) { this.intents = intents; return this; }
        public Builder baseUrl(String baseUrl) { this.baseUrl = baseUrl; return this; }
        public Builder applicationId(String applicationId) { this.applicationId = applicationId; return this; }
        public Builder shardCount(int n) { this.shardCount = n; return this; }

        public Bot build() { return new Bot(this); }
    }
}
